### ISSUE TITLE
CS8509 reported for exhaustive switch expression over single-case union

### ISSUE THREAD

**Issue by eiriktsarpalis** [dotnet/roslyn#83666]
## Summary

A switch expression over a single-case C# union reports CS8509 even though the switch handles the union's only case type. Based on the union proposal, a union switch expression should be exhaustive when all case types have been matched, without requiring a fallback arm.

This appears to affect direct single case-type patterns such as `int i => ...`. Adding a discard arm suppresses the warning, but that defeats union exhaustiveness.

## Minimal repro

```csharp
public union IntUnion(int);

public static class Repro
{
    public static object Deconstruct(IntUnion value) => value switch
    {
        int i => i,
    };
}
```

Compile with `LangVersion=preview`.

## Expected behavior

No CS8509 warning. `IntUnion` has exactly one case type, `int`, and the switch expression covers that case type.

The union proposal says:

> A union type is assumed to be "exhausted" by its case types. This means that a `switch` expression is exhaustive if it handles all of a union's case types.

## Actual behavior

Roslyn reports CS8509 on the switch expression:

```text
CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern '_' is not covered.
```

**Comment by jjonescz** [comment]
I think this is by design. There are two warnings; CS8509 is reported if you are missing one of the known cases, CS8655 is more defensive (since it's possible that union contains other values, e.g., if it comes from a DLL that was updated at runtime) and you can usually just suppress the latter. It's similar to `enum`s (even handling all the `enum` cases doesn't mean there can't be another `int` value in the variable).

@AlekseyTs to confirm.

**Comment by eiriktsarpalis** [comment]
Not sure I entirely follow why this could be by design or why it might be comparable to enums. While it's possible for any union type to return arbitary values, the point of the feature is providing exhaustiveness checking based on a union's declared contract. This should hold for single-case unions, even if it's a trivial case unlikely to have application in practice.

**Comment by jjonescz** [comment]
In other words, I'm saying that if you add `NoWarn=CS8655` to your project, then you will get the behavior you want. CS8509 will warn for missed cases as you expect.

It's comparable to enums like this:

```cs
E e = default;
_ = e switch { E.A => 1 }; // no CS8509 unless you add another case to the enum;
  // but you still get CS8655 unless you handle all integers (but I recommend to suppress CS8655 globally)

enum E { A }
```

**Comment by AlekseyTs** [comment]
I do not see any errors or warnings reported for the repro scenario when the latest state of the feature branch is used. See https://lab.razor.fyi/#nZC_SgNBEMY5IqLbKHmCAZsEJPYGFdEmVpKghZV76ySO7O0eO7NRCfceNlY-hq2Vz2MtsnfmSLBzmx3mz_f9-NRnR6nL4GdBFwPD3ffOnovW6twioEufUmXMLRmIjryDkZOrVPTISX_YDlm0kAFjNTOMsQxeLRQAwPrc5w9oBM7ReMcSopHeUhDm2kbsw9FxUwE_kpj7WqSRSo-cAKUd2q971VBVSimnC-RSG4TJMwsWg3F0QgUOznxRksUwwTAng7xORU4wTNPZqGb4NfsTwdK9wT-B6xpwATOUIVQr-3fE7UGVyP4J1uRYM52KBMqjIBwuNdrWSjjJrbszRS0xIB_Up3yxiSH4wDfbW1_fb68fs92N2-wpe8myHw

However, I do see this behavior in 18.7.0 Insiders [11726.51.main]. I assume the issue has been fixed. Please verify once new feature preview becomes available.
